src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -811,26 +893,43 @@ private ItemSelection GetDefaultsMatch(IReadOnlyList<MatchResult> items, ItemSel
                     }
                 }
 
-                foreach (var defaultText in _snapshotData.Defaults)
-                {
-                    cancellationToken.ThrowIfCancellationRequested();
+                cancellationToken.ThrowIfCancellationRequested();
 
-                    // The range includes all items that are as good of a match as what we initially selected (and in descending order of matching score)
-                    // so we just need to search for the first item that matches the suggested default.
-                    for (var i = 0; i < inferiorItemIndex; ++i)
+                // The range includes all items that are as good of a match as what we initially selected (and in descending order of matching score)
+                // so we just need to search for a preferred item with best default match.
+                var bestPromotedItemSoFar = -1;
+                var bestDefaultIndex = -1;
+                for (var i = 0; i < inferiorItemIndex; ++i)
+                {
+                    var item = matches[i].CompletionItem;
+                    if (item.IsPreferredItem())
                     {
-                        if (items[i].CompletionItem.FilterText == defaultText)
+                        var defaultIndex = _snapshotData.Defaults.IndexOf(matches[i].CompletionItem.FilterText);
+
+                        // This is not a starred item that matches default
+                        if (defaultIndex < 0)
+                            continue;
+
+                        if (bestPromotedItemSoFar < 0 || defaultIndex <= bestDefaultIndex)
                         {
-                            // If user hasn't typed anything, we'd like to hard select the default item.
-                            // This way, they can easily commit the default item which matches what WLC shows.
-                            var selectionHint = _filterText.Length == 0 ? UpdateSelectionHint.Selected : initialSelection.SelectionHint;
-                            return initialSelection with { SelectedItemIndex = i, SelectionHint = selectionHint };
+                            bestPromotedItemSoFar = i;
+                            bestDefaultIndex = defaultIndex;
                         }
+
+                        // We have found the top default item, no need to proceed.
+                        if (bestDefaultIndex == 0)
+                            break;
                     }
                 }
 
                 // Don't change the original selection since there's no match to the defaults provided.
-                return initialSelection;
+                if (bestPromotedItemSoFar < 0)
+                    return initialSelection;
+
+                // If user hasn't typed anything, we'd like to hard select the default item.
+                // This way, they can easily commit the default item which matches what WLC shows.
```

Comment:
what happens if the user has wlc turned off?

Reply:
I assume the list of default would be empty. 
Guess I can add more checks for this to make the code slightly more performant :)